### DISCUSSION TITLE
如何给 target 应用多个 rule

### DISCUSSION THREAD

**Discussion by star-hengxing** [xmake-io/xmake#3282] [Q&A] [answered]
这里想把二进制文件 codegen 到 cpp 文件和头文件，然后编译成静态库或动态库

1. 第一个规则，编译成二进制
2. 第二个规则，把二进制 codgen 到 cpp
3. 编译 cpp 成库

现在我在第二个规则添加
```lua
-- rule2
add_deps("rule1", {order = true})

on_build_files(function (target, sourcebatch, opt)
    print("on_build_files")
end)
```

- 然后发现`on_build_files`并没有运行，不知道为什么（去掉`add_deps`就可以运行）
- 如果上面都没问题，该如何调用`c++.build` rule 进行编译呢

我看文档是这样写
```lua
-- rule2
on_load(function (target)
    local rule = target:rule("c++.build"):clone()
    rule:add("deps", "bin2lib.build", {order = true})
    target:rule_add(rule)
end)
```
但运行的时候输出了
```
❗ error: attempt to index a nil value
```

**Comment by waruqi** [comment, marked as answer]
参考现有 rule 
https://github.com/xmake-io/xmake/blob/master/xmake/rules/lex_yacc/lex/xmake.lua

**Reply by star-hengxing** [reply]
> 参考现有 rule https://github.com/xmake-io/xmake/blob/master/xmake/rules/lex_yacc/lex/xmake.lua

最后一步编译 c++ ，我大概了解了
但第一步和第二步还是有那个问题，我换成`on_buildcmd_file`还是没有执行，这可能是哪里有问题呢，该如何去 debug 出来

**Reply by waruqi** [reply]
set_extensions 匹配对应 files

**Reply by star-hengxing** [reply]
找到原因了，rule2 没有 sourcefiles，所以没有执行，于是我手动用`target:sourcebatches()`在`on_load`的时候`target:add("files", xxx)`，这时候`on_buildcmd_file`就执行了。。
那应该在 rule1 `on_buildcmd_file` 的时候，给 rule2 添加 sourcefiles，不然重复写逻辑有点繁琐，然后我这样做
```lua
-- rule1 on_buildcmd_file
target:add("files", xxx)
print(target:sourcebatches())
```
我看到是加进去 rule2 里了，不过这个时候添加进去，xmake 也不会再执行 rule2 的 on_buildcmd_file 了。。

**Reply by waruqi** [reply]
files只能在load阶段加

**Reply by star-hengxing** [reply]
> files只能在load阶段加

现在又有个问题。。
load 阶段会加载所有 target，但我 rule1 还没有编译出二进制，在 rule2 的`on_load`中 `target:add("files", xxx)`找不到文件就没加进去。。

**Reply by waruqi** [reply]
所以不是让你参考 参考现有 rule https://github.com/xmake-io/xmake/blob/master/xmake/rules/lex_yacc/lex/xmake.lua

 么，不要用 add files 。

**Reply by star-hengxing** [reply]
搞定了，现在完美了，把坑都踩了一圈（其实是在熟悉流程），又可以写一篇文章分享了
